### PR TITLE
Fix Bugzilla 24434 - Casting away const with cast() is not a @safe lv…

### PR DESCRIPTION
**Pull request by ntrel**
…alue

### CODE REVIEW

File: compiler/src/dmd/expressionsem.d [hunk 1]
```diff
@@ -15732,7 +15732,15 @@ private Expression toLvalueImpl(Expression _this, Scope* sc, const(char)* action
             return visit(_this);
         }
         if (_this.isLvalue())
+        {
+            with (_this) // FIXME: avoid new
+            if (mod && !new TypePointer(e1.type).implicitConvTo(new TypePointer(to)))
```
Comment (ntrel) [inline compiler/src/dmd/expressionsem.d:15737]: Is there a way to avoid allocating to get a TypePointer?

Comment (thewilsonator) [reply compiler/src/dmd/expressionsem.d:15737]: `e1.type.pointerTo()` IIRC

**Comment by ntrel** [comment]
```
src/core/internal/array/duplication.d(38):        cast from `ArrElem` to `immutable(ArrElem)` cannot be used as an lvalue in @safe code
src/object.d(3871):        `object.idup!(ArrElem).idup` is declared here
```
That is triggered in `_dupCtfe`:
```d
        U[] res;
        foreach (ref e; a)
            res ~= e; // this line
```
So I'm not sure how to fix that. Also the error should maybe be a deprecation.

**Comment by dkorpel** [comment]
> Also the error should maybe be a deprecation.

Yes, before we have editions, you can use `setUnsafePreview(FeatureState.default_, ...` for that

**Comment by thewilsonator** [comment]
```
../buildkite-ci-build/.dub/packages/libdparse/0.19.4/libdparse/src/dparse/lexer.d(158,23): Deprecation: cast from `const(string)` to `string` cannot be used as an lvalue in @safe code
../buildkite-ci-build/.dub/packages/libdparse/0.19.4/libdparse/src/dparse/parser.d(188,42): Error: template instance `dparse.parser.Parser.attachCommentFromSemicolon!(AliasDeclaration)` error instantiating
../dmd/generated/linux/release/64/dmd -conf= -I../dmd/druntime/import  -w -de -preview=dip1000 -preview=dtorfields -preview=fieldwise -m64 -fPIC -O -release -unittest -version=StdUnittest -c -ofgenerated/linux/release/64/unittest/std/format/package.o std/format/package.d
../buildkite-ci-build/.dub/packages/libdparse/0.19.4/libdparse/src/dparse/lexer.d(150,23): Deprecation: cast from `const(string)` to `string` cannot be used as an lvalue in @safe code
Error /buildkite/builds/geod24-main-1/dlang/dmd/dmd/generated/linux/release/64/dmd failed with exit code 1.
```

Looks like libdparse needs to be updated? (or not compiled with `-de`)

**Comment by ntrel** [comment]
> There is a function called isSafeCast and I think that's where the fix should be located.
>
> On a different note: casting away a qualifier returns an lvalue irrespective of the fact if we are in safe code or not

@RazvanN7 Isn't that actually the same note? isSafeCast doesn't know about lvalues. So I tried making even lvalue QualCast unsafe:
```
std/utf.d(3172): Deprecation: qualifier cast from `immutable(char)` to `char` not allowed in @safe code
...
std/math/algebraic.d(979):        which wouldn't be `@safe` because of:
std/math/algebraic.d(979):        qualifier cast from `const(ulong)` to `ulong` not allowed in @safe code
```
utf.d has lots of `typeof(cast() ...`, which still trigger the deprecation. I'm guessing there's some way of saying to ignore safety inside `typeof`, which would fix those.

The algebraic.d one is `return cast() cast(T) ...` in a private function that returns an rvalue. So things like this which are safe (when T is POD) will be broken if you deprecate/error for rvalues as well.

There is an advantage that this doesn't need the duplication.d change of this pull, which creates a copy instead of `ref` which is probably wrong.

> I don't think this patch solves this case (athough, IMO it should).

It actually does detect that. I'll add it to the test later.

**Comment by ntrel** [comment]
> There is an advantage that this doesn't need the duplication.d change of this pull, which creates a copy instead of ref which is probably wrong.

I undid that change, it is not necessary now I'm doing the check for a qualifier cast correctly.

**Comment by ntrel** [comment]
Now broken - somehow the `mod` field is 0/1 (for mutable/const in the test) as expected for CastExp is visited for semantic analysis, but then by the time `toLvalue` is called, the `mod` field has become 255, and I can't detect the qualifier cast anymore. Maybe the CastExp got replaced with another one, because nowhere modifies `mod`.

**Comment by ntrel** [comment]
As I can't detect a qualifier cast in toLvalue, I decided to also fix the non-qualifier unsafe cast bug too. I found the CatAssignExp semantic code that calls castTo for a safe cast and added a workaround.

**Comment by ntrel** [comment]
buildkite is failing on phobos because libdparse 0.19.4 doesn't have the upstream fix:
```
../buildkite-ci-build/.dub/packages/libdparse/0.19.4/libdparse/src/dparse/lexer.d(158,23): Deprecation: cast from `const(string)` to `string` cannot be used as an lvalue in @safe code
../buildkite-ci-build/.dub/packages/libdparse/0.19.4/libdparse/src/dparse/parser.d(188,42): Error: template instance `dparse.parser.Parser.attachCommentFromSemicolon!(AliasDeclaration)` error instantiating
../buildkite-ci-build/.dub/packages/libdparse/0.19.4/libdparse/src/dparse/lexer.d(150,23): Deprecation: cast from `const(string)` to `string` cannot be used as an lvalue in @safe code
```

**Comment by ntrel** [comment]
Ping. So does this mean we have to wait for libdparse to be released with the fix before this can be merged?

**Comment by RazvanN7** [comment]
Yes, we need to fix libdparse, however, the fix seems trivial.

**Comment by ntrel** [comment]
@RazvanN7 The fix is already merged: https://github.com/dlang-community/libdparse/pull/514

Does the phobos buildkite test require libdparse to be the last released version?

**Comment by thewilsonator** [comment]
> Does the phobos buildkite test require libdparse to be the last released version?

Currently it is testing against libdparse 0.19.4. The current release of libdparse is 0.23.2 released July 2023. 
We will need to tag a new release and then update the version that we test against.

**Review by dkorpel — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@ntrel Phobos seems to continue failing. Does this need a rebase?

**Comment by ntrel** [comment]
@RazvanN7 There's been a [libdparse v0.24.0 release with the fix](https://github.com/dlang-community/libdparse/releases/tag/v0.24.0) but I'm not sure where the file is that specifies 0.19.4 for buildkite - @thewilsonator any idea?

**Comment by dkorpel** [comment]
I think it's https://github.com/dlang/tools/pull/471, and https://github.com/dlang/dlang.org/pull/3815 might also be necessary

**Comment by ntrel** [comment]
buildkite now passing!

**Pull request merged**